c++/src/kj/buffering.c++

```diff
@@ -0,0 +1,248 @@
+#include "buffering.h"
+#include <kj/debug.h>
+
+namespace kj {
+
+namespace _ {
+
+static inline uint lg(uint value) {
+  // Compute floor(log2(value)).
+  //
+  // Undefined for value = 0.
+#if _MSC_VER && !defined(__clang__)
+  unsigned long i;
+  auto found = _BitScanReverse(&i, value);
+  KJ_DASSERT(found);  // !found means value = 0
+  return i;
+#else
+  return sizeof(uint) * 8 - 1 - __builtin_clz(value);
+#endif
+}
+
+FdRangeInputBufferBase::FdRangeInputBufferBase(FdRangeInputBufferOptions options)
+    : options(options),
+      allBytes(heapArray<byte>(options.size)), fds(heapArray<AutoCloseFd>(options.maxFds)) {}
+
+ArrayPtr<byte> FdRangeInputBufferBase::getByteRange() {
+  KJ_SWITCH_ONEOF(state) {
+    KJ_CASE_ONEOF(_, Empty) {
+      return ArrayPtr<byte>(allBytes.begin(), size_t(0));
+    }
+    KJ_CASE_ONEOF(range, FdRange) {
+      return range.bytes;
+    }
+    KJ_CASE_ONEOF(ranges, TwoFdRange) {
+      return ranges.bytes;
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+void FdRangeInputBufferBase::setByteRange(ArrayPtr<byte> bytes) {
+  KJ_SWITCH_ONEOF(state) {
+    KJ_CASE_ONEOF(_, Empty) {
+      KJ_FAIL_ASSERT("buffer is empty");
+    }
+    KJ_CASE_ONEOF(range, FdRange) {
+      range.bytes = bytes;
+    }
+    KJ_CASE_ONEOF(ranges, TwoFdRange) {
+      ranges.bytes = bytes;
+    }
+  }
+}
+
+bool FdRangeInputBufferBase::overflowed() { return allBytes.size() > options.size; }
+
+FdRangeInputBufferBase::ReadParams FdRangeInputBufferBase::prepareForRead(size_t min) {
+  auto nextPowerOf2 = [](size_t n) {
+    // Round up to the next power of 2.
+    size_t result = 1u << (_::lg(n) + 1);
+    KJ_DASSERT(result > n);
+    KJ_DASSERT(result <= n * 2);
+    return result;
+  };
+
+  auto bytes = getByteRange();
+
+  auto spaceNeeded = bytes.size() + min;
+
+  if (spaceNeeded > allBytes.size()) {
+    // We need an overflow buffer. Allocate one (sized to the next power of 2 after spaceNeeded),
+    // copy our buffer into it, and empty the state of our old parent buffer.
+
+    auto overflow = heapArray<byte>(nextPowerOf2(spaceNeeded));
```

Comment:
It looks like you are growing the buffer to accommodate the largest single message it ever receives (plus rounding that up to a power of 2).

I don't think that's the right approach. If a 64MB message is received once you might end up with a 64MB buffer long-term. Though I guess more likely any 64MB message is likely to be retained, so you'd end up transferring away ownership of the buffer (hopefully), though then the power-of-2 round-up is wasted.

I would suggest instead that the buffer size should stay static, and whenever a message larger than the buffer is read, you immediately allocate a new buffer specifically for that message and read into that. Then you don't need to do anything if the message is retained later, as it already owns the buffer.

The benefit of buffered reads is when we can read many _small_ messages at once. It's better to bypass the buffer when reading a very large message.